File: mqtt/c/src/tcp.c
#include <stdio.h>
#include <unistd.h>
#include <arpa/inet.h>

#include "tcp.h"

int connect_tcp(const char *host, int port) {
    int sock;
    struct sockaddr_in server;

    sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock == -1) {
        perror("ERR: ソケットの作成に失敗しました");
        return -1;
    }

    server.sin_family = AF_INET;
    server.sin_port = htons(port);
    inet_pton(AF_INET, host, &server.sin_addr);

    if (connect(sock, (struct sockaddr *)&server, sizeof(server)) < 0) {
        perror("ERR: サーバーへの接続に失敗しました");
        return -1;
    }

    return sock;
}

int disconnect_tcp(int sock) {
    if (close(sock) < 0) {
        perror("ERR: ソケットのクローズに失敗しました");
        return -1;
    }
    return 0;
}

File: mqtt/ruby/lib/tcp.rb
# frozen_string_literal: true

require 'socket'

def connect_tcp(host, port)
  socket = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM, 0)
  sockaddr = Socket.sockaddr_in(port, host)
  socket.connect(sockaddr)
  socket
rescue SocketError => e
  puts 'ERR: ソケットの作成に失敗しました'
  raise e
rescue StandardError => e
  puts 'ERR: サーバーへの接続に失敗しました'
  raise e
end

def disconnect_tcp(socket)
  socket.close
rescue StandardError => e
  puts 'ERR: ソケットのクローズに失敗しました'
  raise e
end
